### PR TITLE
SQL: enforce JDBC driver - ES server version parity

### LINKED ISSUE #38775: SQL: validate the JDBC driver version to be the same as the server
At the moment, only the CLI does the strict version matching between itself and the Elasticsearch server. We should do the same for the JDBC driver (ES server version needs to match the JDBC driver version).

### PR DESCRIPTION
**Pull request by astefan**
This PR enforces the jdbc driver - ES server versions perfect matching.
Due to the low level changes in the http client, one existent test is changed to use a mock web-server. And the version checks tests themselves are built to use the same mocked web-server.

Fixes https://github.com/elastic/elasticsearch/issues/38775.

### CODE REVIEW

File: x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcHttpClient.java [hunk 1]
```diff
@@ -72,16 +74,23 @@ boolean queryClose(String cursor) throws SQLException {
...
 
     private InfoResponse fetchServerInfo() throws SQLException {
         MainResponse mainResponse = httpClient.serverInfo();
         Version version = Version.fromString(mainResponse.getVersion());
-        return new InfoResponse(mainResponse.getClusterName(), version.major, version.minor);
+        return new InfoResponse(mainResponse.getClusterName(), version.major, version.minor, version.revision);
+    }
+    
+    
+    private void checkConnection() throws SQLException {
```
Comment (costin) [inline x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcHttpClient.java]: A more appropriate name would be `checkServerVersion()`.

File: x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/InfoResponse.java [hunk 2]
```diff
@@ -12,10 +12,17 @@
...
+    InfoResponse(String clusterName, byte versionMajor, byte versionMinor, byte revisionVersion) {
         this.cluster = clusterName;
         this.majorVersion = versionMajor;
         this.minorVersion = versionMinor;
+        this.revisionVersion = revisionVersion;
+    }
+
+    @Override
+    public String toString() {
+        return "v" + majorVersion + "." + minorVersion + "." + revisionVersion;
```
Comment (costin) [inline x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/InfoResponse.java]: The toString should contain the cluster name - if the version is needed, make a separate method (`version` or `versionString`).
`return clusterName[major.minor.revision]`

File: x-pack/plugin/sql/qa/security/src/test/java/org/elasticsearch/xpack/sql/qa/security/JdbcSecurityIT.java [hunk 3]
```diff
@@ -230,16 +230,8 @@ public void expectUnknownColumn(String user, String sql, String column) throws E
...
-            expectMatchesAdmin("SELECT * FROM test", user, "SELECT * FROM test");
-            expectMatchesAdmin("SHOW TABLES LIKE 'test'", user, "SHOW TABLES LIKE 'test'");
-            expectMatchesAdmin("DESCRIBE test", user, "DESCRIBE test");
-
-            // But there are a few things that don't work
-            try (Connection es = es(userProperties(user))) {
-                expectUnauthorized("cluster:monitor/main", user, () -> es.getMetaData().getDatabaseMajorVersion());
-                expectUnauthorized("cluster:monitor/main", user, () -> es.getMetaData().getDatabaseMinorVersion());
-            }
+            // Without monitor/main the JDBC driver - ES server version comparison doesn't take place, which fails everything else
```
Comment (costin) [inline x-pack/plugin/sql/qa/security/src/test/java/org/elasticsearch/xpack/sql/qa/security/JdbcSecurityIT.java]: Why remove the test ? it's fairly important in checking security credentials...

**Review by costin — COMMENTED** [review]
Looks over all, left some minor comments. My main concern is around the security test which we should preserve.

File: x-pack/plugin/sql/qa/security/src/test/java/org/elasticsearch/xpack/sql/qa/security/JdbcSecurityIT.java (hunk 3 above)
Comment (astefan) [reply x-pack/plugin/sql/qa/security/src/test/java/org/elasticsearch/xpack/sql/qa/security/JdbcSecurityIT.java]: Every call to any such command will fail with unauthorized exception and at its base will be the connection creation failure, thus the replacement with the connection creation. I will re-add them back.

File: x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcHttpClient.java [hunk 4]
```diff
@@ -72,16 +74,23 @@ boolean queryClose(String cursor) throws SQLException {
...
         return serverInfo;
     }
 
     private InfoResponse fetchServerInfo() throws SQLException {
         MainResponse mainResponse = httpClient.serverInfo();
         Version version = Version.fromString(mainResponse.getVersion());
-        return new InfoResponse(mainResponse.getClusterName(), version.major, version.minor);
+        return new InfoResponse(mainResponse.getClusterName(), version.major, version.minor, version.revision);
+    }
+
```
Comment (matriv) [inline x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcHttpClient.java]: minor: extra empty line

File: x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcHttpClient.java [hunk 5]
```diff
@@ -72,16 +74,23 @@ boolean queryClose(String cursor) throws SQLException {
...
+        return new InfoResponse(mainResponse.getClusterName(), version.major, version.minor, version.revision);
+    }
+    
+    
+    private void checkConnection() throws SQLException {
+        if (serverInfo.majorVersion != Version.CURRENT.major
+                || serverInfo.minorVersion != Version.CURRENT.minor
+                || serverInfo.revisionVersion != Version.CURRENT.revision) {
+            throw new SQLException("This version of the JDBC driver is only compatible with Elasticsearch version " +
+                    Version.CURRENT.toString() + ", but it's configured to connect to a server version " + serverInfo.toString());
```
Comment (matriv) [inline x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcHttpClient.java]: suggestion: maybe `, attempting to connect...` instead

**Review by matriv — COMMENTED** [review]
Nicely done! Left 2 minor comments.

**Review by matriv — APPROVED** [review]
LGTM

**Pull request merged**